http/signature/signature.go

```diff
@@ -0,0 +1,206 @@
+package signature
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	httpx "go.strv.io/net/http"
+)
+
+var (
+	// InputGetError is passed to ErrorHandlerFunc when WrapHandler (or derived) fails in the first step (parsing input)
+	InputGetError = errors.New("parsing input")
+	// InnerHandlerError is passed to ErrorHandlerFunc when WrapHandler (or derived) fails in the second step (inner handler)
+	InnerHandlerError = errors.New("inner handler")
+	// ResponseMarshalError is passed to ErrorHandlerFunc when WrapHandler (or derived) fails in the third step (marshaling response object)
+	ResponseMarshalError = errors.New("marshaling response")
+)
+
+type InputGetterFunc func(r *http.Request, dest any) error
+
+type ResponseMarshalerFunc func(w http.ResponseWriter, r *http.Request, src any) error
+
+type ErrorHandlerFunc func(w http.ResponseWriter, r *http.Request, err error)
+
+type Wrapper struct {
+	inputGetter       InputGetterFunc
+	responseMarshaler ResponseMarshalerFunc
+	errorHandler      ErrorHandlerFunc
+}
+
+func DefaultWrapper() Wrapper {
+	return Wrapper{
+		inputGetter:       UnmarshalRequestBody,
+		responseMarshaler: FixedResponseCodeMarshal(http.StatusOK),
+		errorHandler:      InputGetErrorHandle,
+	}
+}
+
+func (w Wrapper) WithInputGetter(f InputGetterFunc) Wrapper {
+	copied := w
+	copied.inputGetter = f
+	return copied
+}
+
+func (w Wrapper) WithResponseMarshaler(f ResponseMarshalerFunc) Wrapper {
+	copied := w
+	copied.responseMarshaler = f
+	return copied
+}
+
+func (w Wrapper) WithErrorHandler(f ErrorHandlerFunc) Wrapper {
+	copied := w
+	copied.errorHandler = f
+	return copied
+}
+
+func inputErrorWithType(target any, innerError error) error {
+	return fmt.Errorf("%w into type %T: %w", InputGetError, target, innerError)
+}
+
+func responseErrorWithType(src any, innerError error) error {
+	if src == nil {
+		return fmt.Errorf("%w without response object: %w", ResponseMarshalError, innerError)
+	}
+	return fmt.Errorf("%w from type %T: %w", ResponseMarshalError, src, innerError)
+}
+
+func wrapInnerHandlerError(innerError error) error {
+	return fmt.Errorf("%w: %w", InnerHandlerError, innerError)
+}
+
+// WrapHandler enables a handler with signature of second parameter to be used as a http.HandlerFunc.
+// 1. Before calling such inner handler, the http.request is used to get the input parameter of type TInput for the handler, using InputGetterFunc in Wrapper.
+// 2. Then the inner handler is called with such created TInput.
+// 3. If the handler succeeds (returns nil error), The first return value (of type TResponse) is passed to ResponseMarshalerFunc of Wrapper.
+// If any of the above steps returns error, the ErrorHandlerFunc is called with that error.
+func WrapHandler[TInput any, TResponse any](wrapper Wrapper, handler func(http.ResponseWriter, *http.Request, TInput) (TResponse, error)) http.HandlerFunc {
```

Comment:
What if all those `Wrap...` functions were made methods on `Wrapper`? :thinking:

Reply:
cant have generic methods :(

Reply:
That's true :sob: